### PR TITLE
feat: add session impl

### PR DESCRIPTION
**Pull request by darkskygit**

### CODE REVIEW

**Comment by darkskygit** [comment]
* **#6528** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6528?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6527** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6527?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6506** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6525** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6525?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6445** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6445?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6448** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6437** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6437?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6272** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6230** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6254** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6254?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6241** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6229** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @darkskygit and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/backend/server/schema.prisma [hunk 1]
```diff
@@ -30,6 +30,7 @@ model User {
   pagePermissions      WorkspacePageUserPermission[]
   connectedAccounts    ConnectedAccount[]
   sessions             UserSession[]
+  AiSession            AiSession[]
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: camelCase: `aiSessions`

File: packages/backend/server/schema.prisma [hunk 2]
```diff
@@ -96,6 +97,7 @@ model Workspace {
   permissions     WorkspaceUserPermission[]
   pagePermissions WorkspacePageUserPermission[]
   features        WorkspaceFeatures[]
+  AiSession       AiSession[]
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: camelCase: `aiSessions`

File: packages/backend/server/schema.prisma [hunk 3]
```diff
@@ -321,6 +323,8 @@ model Snapshot {
   // but the created time of last seen update that has been merged into snapshot.
   updatedAt   DateTime @map("updated_at") @db.Timestamptz(6)
 
+  AiSession AiSession[]
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: camelCase: `aiSessions`

File: packages/backend/server/schema.prisma [hunk 4]
```diff
@@ -443,6 +447,25 @@ model AiPrompt {
   @@map("ai_prompts")
 }
 
+model AiSession {
+  id          String   @id @default(uuid()) @db.VarChar
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: referenced ids should have strict length of 32

File: packages/backend/server/schema.prisma [hunk 5]
```diff
@@ -443,6 +447,25 @@ model AiPrompt {
...
+  userId      String   @map("user_id") @db.VarChar
+  workspaceId String   @map("workspace_id") @db.VarChar
+  docId       String   @map("doc_id") @db.VarChar
+  promptName  String   @map("prompt_name") @db.VarChar
+  action      Boolean  @db.Boolean
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: what is a boolean action?

File: packages/backend/server/schema.prisma [hunk 6]
```diff
@@ -443,6 +447,25 @@ model AiPrompt {
...
+  model       String   @db.VarChar
+  messages    Json     @db.Json
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: should we put all messages in one session record, which would be too large for a long context services.

File: packages/backend/server/src/plugins/copilot/index.ts [hunk 7]
```diff
@@ -1,11 +1,14 @@
 import { ServerFeature } from '../../core/config';
 import { Plugin } from '../registry';
-import { CopilotPromptService } from './prompt';
-import { assertProvidersConfigs, CopilotProviderService } from './providers';
+import { CopilotController } from './controller';
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/index.ts]: fix CI

File: packages/backend/server/schema.prisma (hunk 5 above)
Comment (darkskygit) [reply packages/backend/server/schema.prisma]: `action` is a bussiness definition, that means the chat is oneshot chat like translate

Comment (forehalo) [reply packages/backend/server/schema.prisma]: better give it a  comment

File: packages/backend/server/schema.prisma [hunk 8]
```diff
@@ -30,7 +30,7 @@ model User {
   pagePermissions      WorkspacePageUserPermission[]
   connectedAccounts    ConnectedAccount[]
   sessions             UserSession[]
-  AiSession            AiSession[]
+  aiSession            AiSession[]
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: 复数

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  private takeMessages(): ChatMessage[] {
+    if (this.state.action) {
+      const messages = this.state.messages;
+      return messages.slice(messages.length - 1);
+    }
+    const ret = [];
+    const messages = [...this.state.messages];
+    messages.reverse();
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts]: why reverse?

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+          model: session.model as TiktokenModel,
+          prompt: await this.prompt.get(session.promptName),
+          messages: messages.success ? messages.data : [],
+        };
+      });
+  }
+
+  private async setState(state: ChatSessionState): Promise<void> {
+    await Promise.all([
+      this.cache.set(`${CHAT_SESSION_KEY}:${state.sessionId}`, state, {
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts]: large key, here is context length, will significantly increase redis latency

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+   * usage:
+   * ``` typescript
+   * {
+   *     // allocate a session, can be reused chat in about 12 hours with same session
+   *     await using session = await session.getOrCreate(sessionId, promptName, model);
+   *     session.push(message);
+   *     copilot.generateText(session.finish(), model);
+   * }
+   * // session will be disposed after the block
+   * @param sessionId session id
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts:191]: signature changed, update the comment

File: packages/backend/server/schema.prisma (hunk 6 above)
Comment (darkskygit) [reply packages/backend/server/schema.prisma]: now split messages as independent table

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 10 above)
Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts]: removed, will optimize query based on splited message table

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 11 above)
Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts:191]: fixed

File: packages/backend/server/schema.prisma [hunk 12]
```diff
@@ -444,8 +443,38 @@ model AiPrompt {
...
-  @@unique([name, idx])
-  @@map("ai_prompts")
+  prompt AiPrompt @relation(fields: [promptId], references: [id], onDelete: Cascade)
+
+  @@unique([id, idx])
+  @@map("ai_prompts_messages")
+}
+
+model AiPrompt {
+  id        String   @id @default(uuid()) @db.VarChar
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: it's safe to use Int

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 9 above)
Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts]: used to calculate message token to keep as much of the chat history as possible

File: packages/backend/server/schema.prisma [hunk 13]
```diff
@@ -444,8 +443,38 @@ model AiPrompt {
   content   String       @db.Text
   createdAt DateTime     @default(now()) @map("created_at") @db.Timestamptz(6)
 
-  @@unique([name, idx])
-  @@map("ai_prompts")
+  prompt AiPrompt @relation(fields: [promptId], references: [id], onDelete: Cascade)
+
+  @@unique([id, idx])
```
Comment (forehalo) [inline packages/backend/server/schema.prisma]: `@@unique([promptId, idx])` and the `id` column is ok to be removed

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 9 above)
Comment (forehalo) [reply packages/backend/server/src/plugins/copilot/session.ts]: iterate messages from end to front will save one reverse turn

Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts]: image

just make bench and found that reverse would indeed cause a significant performance decline.
I have switched to while loop

File: packages/backend/server/src/plugins/copilot/session.ts [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    if (this.state.prompt.action) {
+      const messages = this.state.messages;
+      return messages.slice(messages.length - 1);
+    }
+    const ret = [];
+    const messages = this.state.messages.slice();
+
+    let size = this.state.prompt.tokens;
+    while (messages.length) {
+      const message = messages.pop();
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/session.ts:59]: `pop` is still a array mutation op,

```
for (let i = messages.length - 1; i >= 0; i--)
```

File: packages/backend/server/migrations/20240402100608_ai_prompt_session_metadata/migration.sql [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    "content" TEXT NOT NULL,
+    "created_at" TIMESTAMPTZ(6) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    "updated_at" TIMESTAMPTZ(6) NOT NULL,
+    CONSTRAINT "ai_sessions_messages_pkey" PRIMARY KEY ("id")
+);
+
+-- CreateTable
+CREATE TABLE "ai_sessions_metadata" (
+    "id" VARCHAR NOT NULL,
+    "user_id" VARCHAR NOT NULL,
```
Comment (forehalo) [inline packages/backend/server/migrations/20240402100608_ai_prompt_session_metadata/migration.sql]: VARCHAR(36)

File: packages/backend/server/src/data/migrations/1712068777394-prompts.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+import { prompts } from './utils/prompts';
+
+export class Prompts1712068777394 {
+  // do the migration
+  static async up(db: PrismaClient) {
+    await db.$transaction(async tx => {
+      await Promise.all(
+        prompts.map(prompt =>
+          tx.aiPrompt.create({
```
Comment (forehalo) [inline packages/backend/server/src/data/migrations/1712068777394-prompts.ts:11]: createMany

File: packages/backend/server/src/plugins/copilot/prompt.ts [hunk 17]
```diff
@@ -22,51 +139,74 @@ export class PromptService {
...
+          action: true,
+          model: true,
+          messages: {
+            select: {
+              role: true,
+              content: true,
+              params: true,
+            },
+            orderBy: {
+              createdAt: 'asc',
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/prompt.ts]: `idx: 'asc'`?

File: packages/backend/server/src/plugins/copilot/types.ts [hunk 18]
```diff
@@ -9,6 +15,76 @@ export interface CopilotConfig {
...
+    // dalle don't need to calc the token
+    return undefined;
+  } else {
+    return get_encoding('cl100k_base');
+  }
+}
+
+// ======== ChatMessage ========
+
+export const ChatMessageRole = Array.from(Object.values(AiPromptRole)) as [
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/types.ts]: Object.values returns array already

File: packages/backend/server/src/plugins/copilot/session.ts (hunk 14 above)
Comment (darkskygit) [reply packages/backend/server/src/plugins/copilot/session.ts:59]: `messages.slice()` has copied the array and will not change the original array
according to the previous test results, the while pop combination is slightly faster than for, and the code is clearer:

image

File: packages/backend/server/src/data/migrations/1712068777394-prompts.ts (hunk 16 above)
Comment (darkskygit) [reply packages/backend/server/src/data/migrations/1712068777394-prompts.ts:11]: `createMany` not support craete with connect yet

File: packages/backend/server/src/plugins/copilot/providers/index.ts [hunk 19]
```diff
@@ -87,8 +87,8 @@ export function assertProvidersConfigs(config: Config) {
 }
 
 @Injectable()
-export class CopilotProviderService {
-  private readonly logger = new Logger(CopilotProviderService.name);
+export class ProviderService {
+  private readonly logger = new Logger(ProviderService.name);
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/providers/index.ts]: nonsense name for logger

**Review by forehalo — APPROVED** [review]

**Pull request merged**